vm_insnhelper.c

```diff
@@ -5382,7 +5382,11 @@ vm_opt_ltlt(VALUE recv, VALUE obj)
     }
     else if (RBASIC_CLASS(recv) == rb_cString &&
 	     BASIC_OP_UNREDEFINED_P(BOP_LTLT, STRING_REDEFINED_OP_FLAG)) {
-	return rb_str_concat(recv, obj);
+	if (LIKELY(!SPECIAL_CONST_P(obj) && RBASIC_CLASS(obj) == rb_cString)) {
```

Comment:
```suggestion
	if (LIKELY(RB_TYPE_P(obj, T_STRING))) {
```
`rb_str_concat` calls `rb_str_append` when the argument is not an integer.
And `rb_str_append` just calls `StringValue`, which converts non-`T_STRING` objects.

Reply:
Ah good point, we're ok with subclasses.